### PR TITLE
Attempt to reverse slowdown from hasattr  needed for cached_property

### PR DESCRIPTION
**Pull request by dschult**
As discussed in #5796 there was a slowdown in e.g. `copy` when we changed the base classes to use `cached_property` instead of `property`.  This was due to the need to check whether to reset the cached property during `__init__`.  But it actually doesn't have to happen during `__init__`. So I have moved that check to the places in our codebase where we are replacing or changing `G._adj`/`G._succ`/`G._pred`. The danger is that we change `G._adj` and don't get an updated cached_property.  Within our codebase this is only needed inside `graphviews.py`.  It is hard to know how many users would be affected. If they swap out `G._adj` without resetting the cache property on `G.adj` then they might see some strange results where `G.adj` reports one thing while `G._adj` reports another.

We could avoid all this trouble by computing `G.adj` as an attribute instead of a property.  It would cause creation of the `AdjacencyView` during `__init__` rather than on the first lookup of `G.adj`. So some benchmarks might show a slowdown. Indeed, creating a graph would be slower by the amount of time needed to create the AdjView. Presumably that is more than the time to do a `hasattr` which caused the slowdown we are currently fixing. The tradeoff is between potential for users playing with private attributes to mix things up vs extra work creating each graph object. 

Note: I only really looked at the regression for `copy` assuming it would fix the others. We should check that the others don't have that slowdown after this change.

### CODE REVIEW

**Comment by stefanv** [comment]
This will work, but feels a bit fragile: you'd need to remember to reset the cache whenever modifying certain internal attributes. Is that not perhaps something we'd like to have happen automatically on assignment to those members?

I hope I'm understanding the issue correctly; please disregard if not!

**Comment by dschult** [comment]
Yes -- it would be best to have the cache reset each time someone assigns to private attributes `_adj`, `_succ`, `_pred`.  How might that be done without slowing down other assignments too much?

**Comment by stefanv** [comment]
Could settable properties do the trick (fset for the internal data storage, that clears the cache)? It seems a bit roundabout to have both a property and cached property, but should do the trick.

**Comment by stefanv** [comment]
I like the centralized approach!

**Comment by dschult** [comment]
Yes, the centralized approach is definitely better.  Thanks for the suggestion!!

I'm learning more about __setattr__ too.  
(And making cached_properties that depend on properties that depend on attributes. But I've put aside that for now.)

**Comment by dschult** [comment]
How do I fix a mypy error due to the type of the `_adj` attribute changing when we subclass Graph to DiGraph?

File: networkx/classes/digraph.py [hunk 1]
```diff
@@ -266,6 +285,10 @@ class DiGraph(Graph):
     a dictionary-like object.
     """
 
+    _adj = _CachedPropertyResetterAdjAndSucc()
+    _succ = _CachedPropertyResetterAdjAndSucc()
```
Comment (rossbar) [inline networkx/classes/digraph.py:289]: Realistically I think the best way to get mypy to not complain is just to mark the differences with:

```suggestion
    _adj = _CachedPropertyResetterAdjAndSucc()  # type: ignore
    _succ = _CachedPropertyResetterAdjAndSucc()  # type: ignore
```

I found [this SO post](https://stackoverflow.com/questions/58377106/why-does-mypy-report-incompatible-types-when-child-and-parent-class-both-satisfy) which explains what's happening. This is fine and makes sense *in the case where types are explicitly defined*. IMO it's a bit annoying mypy is inferring typing info from the underlying objects in the cases where we don't provide specific typing information (if we were, I assume we'd be able to get around this error by adding an annotation that allows either of the various `_CachedProperty*` objects).

Either way, I think the best thing to do is just ignore it for now!

**Comment by dschult** [comment]
I think this is ready to be merged...  
I'll try to record my musings during this process in this comment along with timing of various ways to handle the classes. The timings are mostly with Graph objects which are simpler than DiGraph. The main difference is thet DiGraph._adj and DiGraph._succ should be the same object. And we have never enforced that.  This Data Descriptor business allows us to do that fairly simply so, I include that in this PR. I used the same principles as justified by the Graph case, but I have minimal timing of the results. I can probably do that fairly easily soon.

Goals:
1) make G.adj a `cached_property` so we don't construct the `AdjacencyView` every time we use `G.adj`,  
2) provide a centralized way to ensure that when `G._adj` is reset, `G.adj` points to the right object (overwriting `G._adj` is rare, but done sometimes in our code especially with a DiGraph that we want to "reverse"). 

So, we need to connect `G.adj` with `G._adj`. When `G._adj` is "set" we should remove the cached value for `G.adj`. I explored three initial ideas to approach this:  Make `G._adj` a property with a setter that updates `G.adj`. Override the `__setattr__` method on G which is inherited from `object` so that `__setattr__` checks for changes of `G._adj` and resets `G.adj` in that case. The third method is to make `_adj` a [Data Descriptor](https://docs.python.org/3/howto/descriptor.html) (a class object with methods `__get__` or `__set__` or `__del__`.)

After trying these methods, it became clear that properties overwrite the "get" process which in turn adds work to `__getattribute__` when looking up `G._adj`.  Also, overriding `G.__setattr__` slows down every assignment to graph attributes (by a very small amount that probably isn't relevant but seems suboptimal). The resulting solution is to use a Data descriptor with only a `__set__` method (no `__get__` or `__del__`). This avoids interfering with any "get" processes, including `__getattribute__`.  When `G` sets an attribute it checks for a class attribute of that name, and if that attribute has a `__set__` method, then it calls that to perform the "set" operation. I made the homemade "set" operation simply store the value in `G.__dict__` with the name `_adj`.  Thus "get" operations find it as if it wasn't a data descriptor, and when `_adj` is set/reset we can reset the cache on `G.adj`.

Further testing showed that it is faster to hardcode in the name of the attribute rather than store the name on the Data Descriptor itself. Storing the name is pretty nice actually if we had a lot of these I would use it. A method `__set_name__` is called as the Data Descriptor is being created and you can store that and use it to create key names in `G.__dict__` and also to find the name of the `cached_property`.  But the lookup of the name in the Data Descriptor seems to slow down the set operation.  I even tried using `__slots__` on the Data Descriptor but hardcoding the name is better.

The resulting code for the Data Descriptor for `_adj` looks like:
```python
class SetOnlyDescriptorResetCachedPropertyAdj:
    def __set__(self, obj, value):
        od = obj.__dict__
        od["_adj"] = value
        if "adj" in od:
            del od["adj"]
```
It basically sets the entry in `G.__dict__` and then deletes the cached value for `adj` in `G.__dict__`.

This works similarly for `DiGraph._pred` with a different name. The Data Descriptor for `DiGraph._adj` and `DiGraph._succ` are really the same object but stored on two attribute names. "get" should return the same object. And "set' should update the returned object for both attribute names. Similarly, the two `cached_property`s should both be updated whenever the underlying data descriptor is set a new value. So, that Descriptor code is:
```python
class SetOnlyDescriptorResetCachedPropertyAdjAndSucc:
    def __set__(self, obj, value):
        od = obj.__dict__
        od["_adj"] = value
        od["_succ"] = value
        if "adj" in od:
            del od["adj"]
        if "succ" in od:
            del od["succ"]
```

I'll put the timing results in the next comment.

**Comment by dschult** [comment]
The first two tests show that the "get" portion of the Data Descriptors slow things down compared to approaches that only mess with "set" operations.  (Note:  a `property` is a Data Descriptor.)
```
========================================================================
Benchmark: Adding edges: 100000 reps, min of 7 runs
setup:--------------------
elist=[(i,i+3) for i in range(100-3)]
G.add_nodes_from(range(100))
test :--------------------
G.add_edges_from(elist)
========================================================================
CachedGraphProperties                   : 100.0%    8.332e+00
CachedGraphDataDescriptor               : 105.2%    8.764e+00
CachedGraphSetattr                      :  64.9%    5.403e+00
CachedGraphSetOnlyDataDescriptor        :  67.9%    5.657e+00
CachedGraphSetOnlyDataDescriptorTextName:  67.7%    5.641e+00
------------------------------------------------------------------------

========================================================================
Benchmark: Graph._adj: 100000 reps, min of 30 runs
setup:--------------------
NX.add_path(G, range(100))
test :--------------------
G._adj
========================================================================
CachedGraphProperties                   : 100.0%    1.089e-02
CachedGraphDataDescriptor               : 113.0%    1.230e-02
CachedGraphSetattr                      :  29.3%    3.194e-03
CachedGraphSetOnlyDataDescriptor        :  31.0%    3.371e-03
CachedGraphSetOnlyDataDescriptorTextName:  29.9%    3.253e-03
------------------------------------------------------------------------
```
The next test show that the "set" operations are slower for approaches that overwrite the set operation -- though not if you hardcode the name of the attribute.
```
========================================================================
Benchmark: reset Graph._adj: 100000 reps, min of 30 runs
setup:--------------------
NX.add_path(G, range(100))
G.adj
test :--------------------
G._adj={}
========================================================================
CachedGraphProperties                   : 100.0%    1.829e-02
CachedGraphDataDescriptor               :  93.6%    1.712e-02
CachedGraphSetattr                      : 121.2%    2.218e-02
CachedGraphSetOnlyDataDescriptor        : 147.3%    2.694e-02
CachedGraphSetOnlyDataDescriptorTextName:  94.7%    1.732e-02
------------------------------------------------------------------------
```
The next test simply demonstrates that using `__setattr__` does affect setting of other attributes too while the data descriptor methods don't.
```
========================================================================
Benchmark: Graph.foo=4: 100000 reps, min of 30 runs
setup:--------------------
NX.add_path(G, range(100))
test :--------------------
G.foo=4
========================================================================
CachedGraphProperties                   : 100.0%    4.317e-03
CachedGraphDataDescriptor               :  97.7%    4.219e-03
CachedGraphSetattr                      : 425.7%    1.838e-02
CachedGraphSetOnlyDataDescriptor        : 103.0%    4.445e-03
CachedGraphSetOnlyDataDescriptorTextName:  99.3%    4.287e-03
------------------------------------------------------------------------
```
Graph creation is also slower with `__setattr__` overwritten, but this is very suspect timing and comes out slightly different each time I run it.
The next test is copying the graph and that involves creating a new one.
```
========================================================================
Benchmark: create Graph: 100000 reps, min of 30 runs
setup:--------------------
MyGraph=G.__class__
test :--------------------
MyGraph()
========================================================================
CachedGraphProperties                   : 100.0%    1.022e-01
CachedGraphDataDescriptor               :  98.5%    1.006e-01
CachedGraphSetattr                      : 237.5%    2.427e-01
CachedGraphSetOnlyDataDescriptor        : 115.6%    1.181e-01
CachedGraphSetOnlyDataDescriptorTextName: 105.5%    1.078e-01
------------------------------------------------------------------------

========================================================================
Benchmark: Copy Graph: 100000 reps, min of 30 runs
setup:--------------------
elist=[(i,i+3) for i in range(100-3)]
test :--------------------
G.copy()
========================================================================
CachedGraphProperties                   : 100.0%    2.430e-01
CachedGraphDataDescriptor               :  91.8%    2.232e-01
CachedGraphSetattr                      : 143.5%    3.487e-01
CachedGraphSetOnlyDataDescriptor        :  94.9%    2.307e-01
CachedGraphSetOnlyDataDescriptorTextName:  87.6%    2.130e-01
------------------------------------------------------------------------
```
The last test chacks a downstream operation just to check that these small differences do impact the bigger picture operations somewhat.  For reporting neighbors (and also reporting edges, though not shown here) the "get" operations are key, so keeping those fast does help somewhat.
```
========================================================================
Benchmark: reporting neighbors: 1000 reps, min of 7 runs
setup:--------------------
H=NX.binomial_graph(500,0.3,seed=42)
for (u,v) in H.edges():
 G.add_edge(u,v)

test :--------------------
for n in G:
 for nbr in G.neighbors(n):
  pass
========================================================================
CachedGraphProperties                   : 100.0%    9.660e-01
CachedGraphDataDescriptor               : 102.6%    9.914e-01
CachedGraphSetattr                      :  85.5%    8.258e-01
CachedGraphSetOnlyDataDescriptor        :  86.5%    8.360e-01
CachedGraphSetOnlyDataDescriptorTextName:  89.0%    8.601e-01
------------------------------------------------------------------------
```

All of these benchmarks show that we aren't talking about large changes in speed here!! 
Bike-shedding alert!  But I feel better with something this fundamental to make a good choice.

These timings are all done with the `timeit.Timer` builtin object. Code is based on old code that was in the repo for 10 years or more, but removed a few years back since it wasn't being updated or used. [The code may be ugly and obviously a hack](https://gist.github.com/dschult/baa5fd69d81304d4959cee9d3a1443de)  that has lots of redundant lines of code that I copy out or include as I go. But it's an old friend and I'm comfortable with it. :}

**Review by jarrodmillman — APPROVED** [review]

**Comment by stefanv** [comment]
It feels like the current approach abuses Data Descriptors ever so slightly (thanks for teaching me about those, though, didn't know about them!). The timings here are not so different that it should make much of a practical difference, so then I wonder if it wouldn't be better to take a clear, intuitive approach. Why, then, not use the Data Descriptors as they are intended and store values on the DD itself, and add a `__get__` method to reach them (i.e. stop using cached_property, which essentially does the same)?

I see there are some places where `_adj` is explicitly accessed internally; is there a reason, once Data Descriptors are used, to still access that attribute apart from for deletion?

**Comment by dschult** [comment]
I don't understand what you are thinking yet, can you explain more about a clear intuitive approach? Which approach is that?

Without properties or anything fancy, `G.adj` is an object which depends on `G._adj`. That makes it a good candidate to be a property -- and to avoid recreating an identical object every time, make it a cached_property.  But if `G._adj` changes, then we need to replace the cached value presented by `G.adj`.  So `G._adj` should be some sort of Data Descriptor -- maybe a property, but not quite. We only need to impact the set method of the G._adj attribute handling.  So looking at `__getattribute__` (in its python pseudocode equivalent posted in the Python docs DD description) there is a hook specifically put into Python just for this case. If there is a get and a set on a class attribute, that blocks the `__dict__` storage for that attribute on the instance.  But if you only have a set method on the class attribute, the `__dict__` storage is available for lookup in the normal way.

We want the `get` process for `G._adj` to be just like a normal attribute. But the `set` process should be overwritten to update `G.adj`.  That's a DD with a "set" method defined. 

I'm not sure what the property decorator does actually...  I can't figure out how to only get a set method.  For the property function, it seems that if you only provide a "set" method, it doesn't allow any "get" operation at all. So I might be missing something -- properties are confusing. Data descriptors are in my view easier to understand. No magic once the `__getattribute__` process is laid out.

Where can I learn more about properties?

**Comment by stefanv** [comment]
Maybe I should gather one other piece of information before I respond: why do we need adj to refer to _adj. Could we make them one and the same using a descriptor class?

**Comment by dschult** [comment]
G._adj is a dict-of-dict-of-dict.   G.adj is a weird read-only version of that structure. Namely G.adj should be read only for the first and second dict, the third inner most dict can be read/write.  The idea is to make the graph structure immutable via G.adj so people don't accidentally mess up the graph structure by assigning to G.adj[3] or to G.adj[3][4].  We still want them to be able to assign to G.adj[3][4]["color"].

We internally use G._adj to update the data structure -- add edges, etc. But we assume users who assign to G._adj[3] are responsible for knowing what they are doing.

So the G.adj interface is a public facing read-only object (to the next-to-bottom depth of the dict-of-dict-of-dict chain).
The G._adj interface is for internal use and allows direct access at all levels to the dict-of-dict-of-dict structure.
Before nx 2.0 we used a single object G.adj and thus exposed the inner levels of the data structure. We could certainly go back to read/write access on all levels of G.adj. That is much simpler. But susceptible to unintentional breakage.

Could we make G.adj and G._adj a single object with properties?  I'm not sure. I don't see a way to make G.adj[3] read-only while still allowing mutation internally without two objects.  Do you?

**Comment by dschult** [comment]
That's a good start... but now the G._adj set value is a dict that we need to be able to update.

```python
class CachedAdjacencyMatrix:
    _cached_computed_adj = None

    def __set__(self, obj, value):
        print("Caching adjacency matrix")
        self._cached_computed_adj = value.keys()

    def __get__(self, obj, objtype=None):
        return self._cached_computed_adj

class Graph:
    _adj = CachedAdjacencyMatrix()

    @property  # read-only attribute
    def adj(self):
        return self._adj


g = Graph()
g._adj = {1: 1, 2: 2, 3: 3}
print(g.adj)

# This will fail:
# g.adj = {1: 1}

# but so will this 
G._adj[4] = 4
```
How do we change G._adj?

**Comment by stefanv** [comment]
Having to catch modifications to the dictionaries definitely complicates things, but I didn't see code for handling that in the PR. The only way to do that, AFAIK, is to have a custom dictionary in place.  Something like this:

```
class EventOnUpdateDict(dict):
    def __init__(self, d, hook):
        self._hook = hook
        super().__init__(d)

    def __setitem__(self, key, val):
        super().__setitem__(key, val)
        self._hook()


class CachedAdjacencyView:
    def __get__(self, obj, objtype=None):
        if obj._adj_view is None:
            print(f'Recalculating view for {obj._adj}')
            obj._adj_view = str(obj._adj) + '__cached'
        return obj._adj_view


class Graph:
    # Storage for cached view of adjacency matrix
    _adj_view = None

    # Cached view of adjacency matrix, for user consumption
    adj = CachedAdjacencyView()

    @property
    def _adj(self):
        return self._adj_dict

    @_adj.setter
    def _adj(self, value):
        self._adj_dict = EventOnUpdateDict(value, self.refresh_adj_view)
        self.refresh_adj_view()

    def refresh_adj_view(self):
        self._adj_view = None

    def __init__(self):
        # Storage of adjacency matrix
        self._adj_dict = EventOnUpdateDict({}, self.refresh_adj_view)


print('Creating graph')
g = Graph()
print(f'{g.adj=}', '\n')

print('Assigning g._adj')
g._adj = {1: 2, 3: 4, 4: 5}
print(f'{g.adj=}', '\n')

print('Setting value in adjacency storage')
g._adj[1] = 4
print(f'{g.adj=}')
```

But, when this has to be tracked, I think your approach of storing the data on the parent class is the correct thing to do.

**Comment by dschult** [comment]
I think so long as we keep the original value on the Graph object's `__dict__` we can modify the dict keys and values using G._adj[3]=... without another helper class. With a SetOnly data descriptor on the class, i.e. Graph._adj and an actual dict in the instance `__dict__['_adj']`, we can mutate G._adj's contents without using any properties or DDs. And when we reset the whole thing with `G._adj = {}` the DD `__set__` method takes care of clearing the cache on the cached_property `G.adj`.

Here's a version where G._adj is a dict, and G.adj is a dict_keys view of the dict.
```python
class SetOnlyDataDescriptorAdj:
    def __set__(self, obj, value):
        obj.__dict__["_adj"] = value
        if "adj" in obj.__dict__:
            delattr(obj, "adj")
            # or del obj.__dict__["adj"]

class Graph:
    _adj = SetOnlyDataDescriptorAdj()

    def __init__(self):
        # Create the initial storage of the adjacency dict
        self._adj = {}

    @cached_property
    def adj(self):
        return self._adj.keys()  # a view that can't be updated

print('Creating graph')
g = Graph()
print(f'{g.adj=}', '\n')

print('Assigning g._adj')
g._adj = {1: 2, 3: 4, 4: 5}
print(f'{g._adj=}')
print(f'{g.adj=}', '\n')

print('Setting value in adjacency storage')
g._adj[1] = 4
print(f'{g._adj=}')
print(f'{g.adj=}', '\n')

# g.adj[3]=4 raises an exception
```

**Comment by stefanv** [comment]
This would work only if the computed quantity you need is tied to the internal storage dictionary (the example I provided above is for where you needed to make an arbitrary computation on a changed dictionary).

But if that is satisfactory, you can get away only with properties:

```python
from functools import cached_property


class Graph:
    _adj_storage = {}

    @property
    def _adj(self):
        return self._adj_storage

    @_adj.setter
    def _adj(self, value):
        del self.adj
        self._adj_storage = value

    @cached_property
    def adj(self):
        return self._adj.keys()

```

**Comment by stefanv** [comment]
Another spelling of the same thing, which I like slightly less because you have to jump between descriptor and graph object:

```python
from functools import cached_property


class ClearAdjProperty:
    value = {}

    def __set__(self, obj, value):
        self.value = value
        del obj.adj

    def __get__(self, obj, objclass=None):
        return self.value


class Graph:
    @cached_property
    def adj(self):
        return self._adj.keys()

    _adj = ClearAdjCacheProperty()
```

**Comment by stefanv** [comment]
Isn't that what the second version above gives? And then there's no messing around with `__dict__`.  I'm always careful of cleverness, since I usually don't understand my own `__`-cleverness six months on :)

**Comment by dschult** [comment]
I think the second version sets a dict on the class ClearAdjCachedSet and then uses it for every graph that is created. I'm sure there is some way to get it to be different for each instance of ClearAdjCachedSet.  But the head-bending continues! 

We only create one Data Descriptor for the Graph class...  not for each instance.  So I think storing the empty dictionary on the Data Descriptor cannot work. 

I think the simple approach is: We store methods on the class and data on the instance. The DD adjusts the methods provided by the class. We should store the data itself on the object -- which I think means putting it into `obj.__dict__`. 

If it is better to use `vars(obj)` instead of `obj.__dict__` that works just fine. But I don't feel I understand `vars(obj)` yet.

**Comment by stefanv** [comment]
You're absolutely right! Thanks for explaining that nuance.

**Comment by jarrodmillman** [comment]
@stefanv Is this ready to merge?

**Comment by stefanv** [comment]
Very nicely done, thank you @dschult!

**Pull request merged**